### ISSUE TITLE
Enhancement: "$TYPE" format keyword

### ISSUE THREAD

**Issue by Naheulf** [dail8859/DoxyIt#23]
It would be nice if you add support for the "$TYPE" format keyword.

For "C", "C++", "Java" and "C#", types are part of the language and can be extracted following rule of each language.
For "Python" and "PHP" types should be extracted following respective types hints: 
- For python https://docs.python.org/3/library/typing.html
- For PHP [functions arguments](https://www.php.net/manual/en/functions.arguments.php) and [functions returning values](https://www.php.net/manual/en/functions.returning-values.php)
- For "JavaScript" /* no comment */

**Comment by dail8859** [comment]
I agree something like this would be useful, the problem it is very difficult to do correctly. Even now the code to parse out function names and parameters only works "well enough" most of the time.

You can quickly get into cases of:

```py
def async_query(on_success: Callable[[int], None], on_error: Callable[[int, Exception], None]) -> None:
```

Which I haven't tried but probably fails to correctly pull out the two parameters currently.

To do it correctly would either involve complex regexes, string manipulation, and corner cases...or else something like a [PEG](https://en.wikipedia.org/wiki/Parsing_expression_grammar) which probably involves more time than I am willing to put into it. But someone else is very welcome to try to tackle this.

**Comment by WinterSilence** [comment]
@dail8859 
> Note that for PHP one can also specify the type (or types if you separate them with a pipe symbol) which are allowed for a parameter (as this is not part of the definition). The syntax is the same as for the phpDocumentor

http://www.doxygen.nl/manual/commands.html#cmdparam

**Comment by dail8859** [comment]
@WinterSilence Yes that adds yet another level of complexity which makes this feature that much more difficult to do correctly.

**Comment by WinterSilence** [comment]
@dail8859 what's problem? explode params by `,` adn parse params: `([^$]*)\$([^\s=]+)`
Also https://github.com/dail8859/DoxyIt/blob/master/src/Parsers/C_Parser.cpp#L24 failed when default value is array `[]` or `array()`, `example.php`:
~~~
Class Test {
  public function foo(array $options = []): string {
     return implode(', ', $options);
  }
}

**Comment by dail8859** [comment]
By all means please create a regex the properly parses out all the different parts of a function/method definition and I'll gladly incorporate it. My suspicion is it is easy to handle the simple cases but only using a regex alone wont be sufficient to handle complex cases.

**Comment by WinterSilence** [comment]
@dail8859 `/([^\$]*)\$([\w]+)/i` cover all cases for php. why u dont use  Doxgen for parsing?

**Comment by dail8859** [comment]
> why u dont use Doxgen for parsing?

I'm not sure what you mean by this.

**Comment by WinterSilence** [comment]
@dail8859 
1. I mean why you use part of code Doxygen? you can call it as application/library.
2. Replace [`(\\$?\\w+|\\.\\.\\.)(\\s*=\\s*[\\\"\\w\\.']+)?\\s*[,)]`](https://github.com/dail8859/DoxyIt/blob/master/src/Parsers/C_Parser.cpp#L24) to `(\\$?\\w+|\\.\\.\\.)(\\s*=\\s*[^,)]+)?\\s*[,)]` for cover more default values(`[]`/`['a' => 1, 'b' => "foo"]`, `Class::CONSTANT`, `'text'`/`"text"`).
3. need something like [input_filter](https://www.doxygen.nl/manual/config.html#cfg_input_filter) to support [filters](https://github.com/AbcAeffchen/doxygen-php-filters)
